### DISCUSSION TITLE
GraphQL - Get all commits by all users in all orgs with filtered date

### DISCUSSION THREAD

**Discussion by udayabaski** [community/community#24590] [Apps, API and Webhooks]
<p>Hi,</p>
<p>  I have the following query which returns all the commit history for the given org. </p>
<p><strong>Questions 1:-</strong></p>
<ol>
<li>
<p>Not sure, How to filter commits history by date</p>
</li>
<li>
<p>Here, I pass orgName. How to get across all orgs?</p>
<p>{<br>
search(query: “org:test”, type: REPOSITORY, last: 100) {<br>
edges {<br>
node {<br>
… on Repository {<br>
name<br>
defaultBranchRef {<br>
name<br>
target {<br>
… on Commit {<br>
history(first: 100) {<br>
edges {<br>
node {<br>
… on Commit {<br>
committedDate<br>
additions<br>
author {<br>
name<br>
email<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}</p>
</li>
</ol>

**Comment by lee-dohm** [comment, marked as answer]
<p>You can filter commit history by date using the <a href="https://developer.github.com/v4/object/commit/" rel="nofollow noopener"><code>since</code> or <code>until</code> parameter to the <code>history</code> connection</a> on <code>Commit</code>:</p>
<pre><code>{
  search(query: "org:test", type: REPOSITORY, last: 100) {
    nodes {
      ... on Repository {
        name
        defaultBranchRef {
          name
          target {
            ... on Commit {
              history(first: 100, since: "2013-07-11T00:00:00") {
                totalCount
                nodes {
                  ... on Commit {
                    committedDate
                    additions
                    author {
                      name
                      email
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>As for getting all orgs, there isn’t currently a way to do that via the GraphQL API because the <code>search</code> connection requires <em>something</em> in the search query. There’s also no top-level connection that allows you to list all users, orgs, or repositories.</p>
<p>I hope that helps!</p>

**Reply by maciekbanas** [reply]
Still that does not give you all commits, but only 100 per repo. Or am I wrong?

**Reply by IAmHughes** [reply]
You would want to pagenate on them after this.

**Reply by maciekbanas** [reply]
You mean with external wrapper on commits - with the use of endCursor and nextPage? But I wonder how to do it if I have repos above (sorry, I'm new to GraphQL). I mean, I can pagenate commits when I have query for one repository, like that:
```
{
  repository(name: "Test", owner: "Owner") {
    defaultBranchRef {
      target {
        ... on Commit {
          history(first: 100) {
            pageInfo {
              hasNextPage
              endCursor
            }
            edges {
              node {
                ... on Commit {
                  id
                  committedDate
                  author {
                    name
                  }
                  additions
                  deletions
                  changedFilesIfAvailable
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Reply by IAmHughes** [reply]
It's not super clean but I have some examples here that was for a project I worked on when I found this discussion: https://github.com/IAmHughes/github-enterprise-user-commit-history

**Reply by maciekbanas** [reply]
Yeah, thanks for that. So as far as I see you do it the way, I described - you first iterate over repos and then over commits, so you handle pagination for them individually (these are separate queries for repos and commits: https://github.com/IAmHughes/github-enterprise-user-commit-history/blob/main/user-commit-history-report.js). 
So, rather your example is the answer to the problem, not the above 'search' query, where you can not handle pagination of both repos and commits in one query.

**Comment by AshantaP** [comment]
<p>How collect all commits from all organizations with all repos for all the users ?<br>
instead of “org:test”, How can we get all commit details for all the orgs ?</p>

**Comment by IAmHughes** [comment]
<p>Hey <a class="mention" href="/lee-dohm">@lee-dohm</a> - now that there’s a concept of an Enterprise object with 1:many organizations, can this be improved?</p>
<p>We are looking for a way to query all commits made by a member of my organization. I need to find out as much commit info as possible.</p>
<p>It would be…</p>
<p>For a given user, find all commits made across every repository in every organization over the last 12 months.</p>
<p>Then for each commit, any data on the commit itself (<code>sha</code>, and if possible, the actual changes made themselves, i.e. maybe the filenames and we’d output the url to see the diff)</p>
